File: 一些问题/8queen.js
/*

八皇后

*/

/*** N皇后 */
class NQueen {

    /*** 方案数目 */
    #num = 0;
    /*** 保存皇后坐标位置的递归数组 */
    #qp = new Array();

    /**
     * 判断当前皇后坐标是否可行
     * @param p 当前皇后坐标
     * @returns bool值，是/否
     */
    #isqpok = function (p) {
        for (let e of this.#qp)
            if (e[0] === p[0] || e[1] === p[1] || Math.abs(e[0] - p[0]) === Math.abs(e[1] - p[1]))
                return false;
        return true;
    }

    /**
     * 内部递归计算
     * @param n 皇后数量
     * @param r 当前行数
     */
    #nq = function (n, r = 0) {
        if (r >= n) {
            this.#num++;
            return;
        }

        for (let i = 0; i < n; i++) {
            let p = [i, r];
            if (this.#isqpok(p)) {
                this.#qp.push(p);
                this.#nq(n, r + 1);
                this.#qp.pop();
            }
        }
    }

    /**
     * n皇后计算
     * @param qn 皇后数量
     * @returns 方案数
     */
    nQueen(qn) {
        if (qn <= 0 || typeof (qn) != "number")
            return 0;
        qn = parseInt(qn);
        this.#num = 0;
        this.#qp.length = 0;
        this.#nq(qn);
        return this.#num;
    }
}

function main() {
    let nqueen = new NQueen();
    console.log(nqueen.nQueen(8));
    console.log(nqueen.nQueen(10));
}

main();

File: 一些问题/8queen.py
'''

八皇后

'''

from typing import Tuple


class NQueen:
    '''
    N皇后
    '''
    def __init__(self):
        self.qpos = []
        self.num = 0

    def qpjud(self, qp: Tuple[int]) -> bool:
        for p in self.qpos:
            if p[0] == qp[0] or p[1] == qp[1] \
               or abs(p[0] - qp[0]) == abs(p[1] - qp[1]):
                return False
        return True

    def calc(self, n: int, r: int = 0):
        if r >= n:
            self.num += 1
            return
        for i in range(n):
            p = (i, r)
            if self.qpjud(p):
                self.qpos.append(p)
                self.calc(n, r + 1)
                self.qpos.pop()

    def nqueen(self, n: int) -> int:
        if n <= 0:
            return 0
        self.qpos.clear()
        self.num = 0
        self.calc(n)
        return self.num


if __name__ == "__main__":
    nqueen = NQueen()
    print(nqueen.nqueen(8))
    print(nqueen.nqueen(10))
